### PR TITLE
Align error-constructor-expr with spec

### LINKED ISSUE #26341: Align the error constructor with SwanLake spec version
**Description:**
After the removal of functional constructors, the current implementation is no longer aligned with the one specified in the spec.

New error constructor is on the grammar:
```
error-constructor-expr := error [error-type-reference] ( error-arg-list )
error-arg-list := positional-arg [, positional-arg] (, named-arg)*
```

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #27502: Ballerina parser crashes due to missing error message, in error message resource bundle
**Description:**
Execute following:
```ballerina
listener Listener|error lsn = new ();
```

Compiler crashes with following log:
```
[2020-12-15 17:11:37,134] SEVERE {b7a.log.crash} - Can't find resource for bundle java.util.PropertyResourceBundle, key error.missing.positional.arg.in.error.constructor 
java.util.MissingResourceException: Can't find resource for bundle java.util.PropertyResourceBundle, key error.missing.positional.arg.in.error.constructor
	at java.base/java.util.ResourceBundle.getObject(Unknown Source)
	at java.base/java.util.ResourceBundle.getString(Unknown Source)
	at io.ballerina.compiler.internal.diagnostics.DiagnosticMessageHelper.getDiagnosticMessage(DiagnosticMessageHelper.java:40)
	at io.ballerina.compiler.internal.diagnostics.SyntaxDiagnostic.message(SyntaxDiagnostic.java:69)
	at org.wso2.ballerinalang.compiler.diagnostic.DiagnosticComparator.compare(DiagnosticComparator.java:71)
	at org.wso2.ballerinalang.compiler.diagnostic.DiagnosticComparator.compare(DiagnosticComparator.java:38)
	at java.base/java.util.TreeMap.put(Unknown Source)
	at java.base/java.util.TreeSet.add(Unknown Source)
	at org.wso2.ballerinalang.compiler.tree.BLangPackage.addDiagnostic(BLangPackage.java:277)
	at org.wso2.ballerinalang.compiler.diagnostic.BLangDiagnosticLog.storeDiagnosticInModule(BLangDiagnosticLog.java:208)
	at org.wso2.ballerinalang.compiler.diagnostic.BLangDiagnosticLog.logDiagnostic(BLangDiagnosticLog.java:169)
	at io.ballerina.projects.DocumentContext.reportSyntaxDiagnostics(DocumentContext.java:194)
	at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:108)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:328)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleCompilationState$1.generatePlatformSpecificCode(ModuleCompilationState.java:53)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:299)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:126)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:116)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:97)
	at java.base/java.util.HashMap.computeIfAbsent(Unknown Source)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:122)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:96)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:153)
	at java.base/java.util.Optional.ifPresent(Unknown Source)
	at org.ballerinalang.tool.Main.main(Main.java:57)
 

```
**Steps to reproduce:**

**Affected Versions:**
SLP7.

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fixes #26341 
Fixes #27502

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by MaryamZi** [comment]
Shall we change the base to https://github.com/ballerina-platform/ballerina-lang/tree/lang-sl-alpha?

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/errorconstructorexpr/ErrorConstructorExprTest.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    @Test
+    public void testErrorConstructorExpr8() {
+        BRunUtil.invoke(result, "testErrorConstructorExpr8");
+    }
+
+    @Test
+    public void testErrorConstructorExpr9() {
+        BRunUtil.invoke(result, "testErrorConstructorExpr9");
+    }
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/errorconstructorexpr/ErrorConstructorExprTest.java:88]: Shall we use a data provider instead?

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+function testMissingParamsNegative() {
+    error e1 = error(); // missing arg within parenthesis
+    error e2 = error(2); // expected 'string', found 'int'
+    error e3 = error(error("Message")); // expected 'string', found 'error'
+    error e4 = error("M1", "M2"); // expected 'error?', found 'string'
+    error e5 = error("M", error("M1"), "c = 2000"); // additional positional arg in error constructor
+
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal:23]: Extra new line.

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+function testMissingParamsNegative() {
+    error e1 = error(); // missing arg within parenthesis
+    error e2 = error(2); // expected 'string', found 'int'
+    error e3 = error(error("Message")); // expected 'string', found 'error'
+    error e4 = error("M1", "M2"); // expected 'error?', found 'string'
+    error e5 = error("M", error("M1"), "c = 2000"); // additional positional arg in error constructor
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal:22]: Can we add similar tests for error constructors with type refs too?

Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal:24]: Added

File: tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal (hunk 2 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/expressions/errorconstructorexpr/error-constructor-expr-negative.bal:25]: Fixed

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/errorconstructorexpr/ErrorConstructorExprTest.java (hunk 1 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/expressions/errorconstructorexpr/ErrorConstructorExprTest.java]: Added

File: compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties [hunk 4]
```diff
@@ -287,3 +287,4 @@ error.rest.arg.in.error.constructor=rest arg in error constructor
 error.additional.positional.arg.in.error.constructor=additional positional arg in error constructor
 error.missing.arg.in.error.constructor=missing arg in error constructor
 error.missing.arg.within.parenthesis=missing arg within parenthesis
+error.missing.positional.arg.in.error.constructor=missing positional arg in error constructor
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-parser/src/main/resources/syntax_diagnostic_message.properties:289]: Removed the previous error

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 5]
```diff
@@ -5002,123 +5068,42 @@ private boolean isFunctionPointer(BSymbol funcSymbol) {
...
-
-        if (iExpr.argExprs.isEmpty() && !iExpr.requiredArgs.isEmpty()) {
-            // This is a special condition that occur due to typechecking from desugar phase.
-            // This only happen for invocations with in object init functions.
-            // It's safe to exit from here as this already has being typechecked.
-            resultType = iExpr.type;
-            return;
-        }
-        if (iExpr.argExprs.isEmpty()) {
-            dlog.error(iExpr.pos, DiagnosticErrorCode.MISSING_REQUIRED_ARG_ERROR_MESSAGE);
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:5026]: Removed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java [hunk 6]
```diff
@@ -1619,10 +1630,6 @@ private void analyzeInvocation(BLangInvocation invocationExpr) {
         }
 
         BInvokableSymbol symbol = (BInvokableSymbol) invocationExpr.symbol;
-        if (symbol == null || symbol.getKind() == SymbolKind.ERROR_CONSTRUCTOR) {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsolationAnalyzer.java:1622]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 7]
```diff
@@ -2647,6 +2648,13 @@ public void visit(BLangInvocation invocationExpr) {
         }
     }
 
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
+        analyzeExprs(errorConstructorExpr.positionalArgs);
+        if (errorConstructorExpr.namedArgs.size() > 0) {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 8]
```diff
@@ -809,6 +810,15 @@ public void visit(BLangInvocation iExpr) {
         result = iExpr;
     }
 
+    @Override
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
+        errorConstructorExpr.positionalArgs = rewriteExprs(errorConstructorExpr.positionalArgs);
+        for (BLangNamedArgsExpression namedArgExpr : errorConstructorExpr.namedArgs) {
+            namedArgExpr.expr = rewriteExpr(namedArgExpr.expr);
+        }
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java [hunk 9]
```diff
@@ -809,6 +810,15 @@ public void visit(BLangInvocation iExpr) {
         result = iExpr;
     }
 
+    @Override
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
+        errorConstructorExpr.positionalArgs = rewriteExprs(errorConstructorExpr.positionalArgs);
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java]: Yes, fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java [hunk 10]
```diff
@@ -1343,27 +1339,26 @@ private void createWait(BLangWaitExpr waitExpr) {
...
+        BIRVariableDcl tempVarError = new BIRVariableDcl(errorConstructorExpr.type,
                 this.env.nextLocalVarId(names), VarScope.FUNCTION, VarKind.TEMP);
 
         this.env.enclFunc.localVars.add(tempVarError);
         BIROperand lhsOp = new BIROperand(tempVarError);
 
-        // visit message, cause and detail expressions
         this.env.targetOperand = lhsOp;
-        invocationExpr.requiredArgs.get(0).accept(this);
+        errorConstructorExpr.positionalArgs.get(0).accept(this);
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/BIRGen.java]: Changed

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+ *  under the License.
+ */
+package org.ballerinalang.model.tree.expressions;
+
+/**
+ * Represent error-constructor-expr.
+ *
+ * @since 2.0.0
+ */
+public interface ErrorConstructorExpressionNode extends ExpressionNode {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java:25]: Added

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java [hunk 12]
```diff
@@ -380,6 +382,17 @@ public void visit(BLangInvocation invocationNode) {
         // eg: string modifiedStr = sampleStr.replace("hello", "Hello").<cursor>toLower();
     }
 
+    @Override
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
+        for (BLangExpression positionalArg : errorConstructorExpr.positionalArgs) {
+            this.acceptNode(positionalArg, this.symbolEnv);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java:388]: Other visit implementations seem to set the scope before calling `acceptNode`, do we need to do the same here?

@nadeeshaan

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+    void addPositionalArg(ExpressionNode positionalArg);
+
+    void addNamedArgArg(NamedArgNode namedArg);
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java:29]: Shall we add getters instead?

Arg is repeated in the second method btw.

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java [hunk 14]
```diff
@@ -93,7 +93,6 @@
     POSITIONAL_ARG_DEFINED_AFTER_NAMED_ARG("BCE2061", "positional.arg.defined.after.named.arg"),
     REST_ARG_DEFINED_AFTER_NAMED_ARG("BCE2062", "rest.arg.defined.after.named.arg"),
     MISSING_REQUIRED_PARAMETER("BCE2063", "missing.required.parameter"),
-    MISSING_REQUIRED_ARG_ERROR_MESSAGE("BCE2064", "missing.required.parameter.error.message"),
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:96]: Shall we update the subsequent codes?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java [hunk 15]
```diff
@@ -8123,15 +8097,19 @@ private void createLHSSafeNavigation(List<BLangStatement> stmts, BLangExpression
...
-                errorCtorInvocation.symbol = symTable.errorConstructor;
+                BLangErrorConstructorExpr errorConstructorExpr =
+                        (BLangErrorConstructorExpr) TreeBuilder.createErrorConstructorExpressionNode();
+                BSymbol symbol = symResolver.lookupMainSpaceSymbolInPackage(errorConstructorExpr.pos, env,
+                        names.fromString(""), names.fromString("error"));
+                errorConstructorExpr.type = symbol.type;
+                errorConstructorExpr.pos = expr.pos;
+                List<BLangExpression> positionalArgs = new ArrayList<>();
+                positionalArgs.add(createStringLiteral(errorConstructorExpr.pos,
+                        "{" + RuntimeConstants.MAP_LANG_LIB + "}InvalidKey"));
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java:8108]: Is the arg correct? Shouldn't we use `literal` here?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java [hunk 16]
```diff
@@ -1994,8 +1995,24 @@ public BLangNode transform(FunctionCallExpressionNode functionCallNode) {
...
+        BLangErrorConstructorExpr errorConstructorExpr =
+                (BLangErrorConstructorExpr) TreeBuilder.createErrorConstructorExpressionNode();
+        errorConstructorExpr.pos = getPosition(errorConstructorExprNode);
+        if (errorConstructorExprNode.typeReference().isPresent()) {
+            errorConstructorExpr.errorTypeRef =
+                    (BLangUserDefinedType) createTypeNode(errorConstructorExprNode.typeReference().get());
+        }
+
+        List<BLangExpression> positionalArgs = errorConstructorExpr.positionalArgs;
+        List<BLangNamedArgsExpression> namedArgs = errorConstructorExpr.namedArgs;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java:2007]: Wouldn't it be clearer if we define two new lists here, populate them in the for loop, and then set

```java
        errorConstructorExpr.positionalArgs = positionalArgs;
        errorConstructorExpr.namedArgs = namedArgs;
```

before returning `errorConstructorExpr`?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java [hunk 17]
```diff
@@ -2647,6 +2648,13 @@ public void visit(BLangInvocation invocationExpr) {
         }
     }
 
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
+        analyzeExprs(errorConstructorExpr.positionalArgs);
+        if (errorConstructorExpr.namedArgs.isEmpty()) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java:2653]: Shouldn't this be

```suggestion
        if (!errorConstructorExpr.namedArgs.isEmpty()) {
```

Please check if we can add a test for this too.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java [hunk 18]
```diff
@@ -2674,6 +2674,78 @@ public void visit(BLangInvocation iExpr) {
...
+                return;
+            }
+
+            if (!types.isAssignable(recordType, targetErrorDetailRec)) {
+                reportErrorDetailMissmatchError(errorConstructorExpr, targetErrorDetailRec, recordType);
+                resultType = symTable.semanticError;
+                return;
+            }
+        } else {
+            resultType = symTable.semanticError;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2743]: Shouldn't we return here? Or it gets overridden in L2746?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * Implementation of error-constructor-expr.
+ *
+ * @since 2.0.0
+ */
+public class BLangErrorConstructorExpr extends BLangExpression implements ErrorConstructorExpressionNode {
+    public BLangUserDefinedType errorTypeRef;
+    public List<BLangExpression> positionalArgs = new ArrayList<>();
+    public List<BLangNamedArgsExpression> namedArgs = new ArrayList<>();
+    // This is added to store namedArgs and it is used only at desugar.
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java:39]: Store the detail mapping created from the named args, right?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+        return NodeKind.ERROR_CONSTRUCTOR_EXPRESSION;
+    }
+
+    @Override
+    public void addPositionalArg(ExpressionNode positionalArg) {
+        positionalArgs.add((BLangExpression) positionalArg);
+    }
+
+    @Override
+    public void addNamedArgArg(NamedArgNode namedArg) {
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java:58]: Extra Arg suffix.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 21]
```diff
@@ -1027,6 +1028,11 @@ public void visit(BLangInvocation invocationExpr) {
         }
     }
 
+    @Override
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
```
[the remark this answers is not in the document]
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:1032]: Added

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java (hunk 13 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/model/tree/expressions/ErrorConstructorExpressionNode.java]: Added

File: compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java (hunk 14 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/ballerinalang/util/diagnostic/DiagnosticErrorCode.java:96]: Updated

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java (hunk 15 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java]: yes. fixed.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java (hunk 16 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java (hunk 17 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/CodeAnalyzer.java]: Added

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java (hunk 18 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java:2743]: Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java (hunk 19 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java]: Yes, fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java (hunk 20 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/tree/expressions/BLangErrorConstructorExpr.java]: Removed

**Review by MaryamZi — APPROVED** [review]
@KavinduZoysa, let's follow-up on verifying the changes to `EnvironmentResolver.java` and `TaintAnalyzer.java` ASAP, and send a separate PR with changes, if required.

File: compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java (hunk 12 above)
Comment (nadeeshaan) [reply compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/EnvironmentResolver.java:388]: Doesn't it handle within the expression itself? I wonder whether the logic here is to just route to the expression visitor and handle the scoping there?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java [hunk 22]
```diff
@@ -1196,12 +1197,28 @@ public void visit(BLangInvocation invocationExpr) {
     }
 
     @Override
-    public void visit(BLangInvocation.BLangActionInvocation actionInvocation) {
-        this.visit((BLangInvocation) actionInvocation);
+    public void visit(BLangErrorConstructorExpr errorConstructorExpr) {
```
[the remark this answers is not in the document]
Comment (rdhananjaya) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TaintAnalyzer.java:1200]: Yes this should be enough. If any of the arguments to an error constructor is tainted, then the constructed error is tainted.

We have a test case covering this `TaintedStatusPropagationTest.testErrorNegative`.
@KavinduZoysa shall we add another test covering an error constructor with a type.

**Pull request merged**